src/TypeChecker/Inference.hs

```diff
@@ -0,0 +1,521 @@
+      in constructorsToEnv tail typeName params env'
+  where 
+    schemeNoParams = Scheme [] (foldr FunctionType 
+                                      (NonPrimType typeName) 
+                                      types)
+    schemeParam = Scheme params 
+                         (foldr FunctionType 
+                                (ParamType typeName (map TypeVariable params)) types)
+      
+
+typecheckAlias :: TAlias -> Environment -> Either String Environment
+typecheckAlias (TAlias (TypeName name) params kind alType _ ) env =
+  -- TODO: check if params are in type / check type correctness?
+  --        or is it done in parsing? 
+  -- TODO: rework
+  Right $ extendEnv env ((TypeVar name), Scheme params alType)
+
+
+typecheckTopLevelLet :: Definition a -> Environment -> Either String Environment
+typecheckTopLevelLet def env = case letPattern def of
+  (VarPattern (Identifier id) pos _ ) -> 
+    let res = inferLetDefExpr (letExpr def) env
+    in case res of
+        Left err -> Left err
+        Right t  -> 
+          let resUni = unifyWithTypeSig t (letTypeSig def)
+              vars = (Set.toList (freeTypeVar t))
+          in case resUni of
+              Left err -> Left err
+              Right _ -> Right (extendEnv env (TypeVar id, Scheme vars t))
+  p@(TuplePattern _ pos _ ) -> 
+    let resPattern = inferLetDefComplexPattern p def env
+    in case resPattern of
+        Left err -> Left err
+        Right (env', exprType) -> 
+          let resUni = unifyWithTypeSig exprType (letTypeSig def)
+          in case resUni of
+              Left err -> Left err
+              Right _ -> Right env' 
+  p@(ConstructorPattern _ _ pos _ ) -> 
+    let resPattern = inferLetDefComplexPattern p def env
+    in case resPattern of
+        Left err -> Left err
+        Right (env', exprType) ->
+          let resUni = unifyWithTypeSig exprType (letTypeSig def)
+          in case resUni of
+              Left err -> Left err
+              Right _ ->  Right env' 
+  (NamedPattern (Identifier id) pattern pos _ ) ->
+    let resPattern = inferLetDefComplexPattern pattern def env
+    in case resPattern of
+      Left err -> Left err
+      Right (env', exprType) -> 
+        let resUni = unifyWithTypeSig exprType (letTypeSig def)
+            vars = (Set.toList (freeTypeVar exprType))
+          in case resUni of
+            Left err -> Left err
+            Right _ -> 
+              Right $ extendEnv env' (TypeVar id, Scheme vars exprType)
+
+  _ -> Left $ "Error: Illegal operation"
+  where
+    unifyWithTypeSig :: Type -> Maybe TypeSig -> Either String Type
+    unifyWithTypeSig t typeSig = 
+      case typeSig of
+        Nothing -> Right t
+        Just (TypeSig _ ts) ->
+          let (res, _) = runInference $ unify t ts
+          in case res of
+              Left _ -> Left $ "Error: Type of the expression: " 
+                            ++ (T.unpack (prettyPrint 0 t))
```

Comment:
It's a bad idea to use the prettyPrinter to construct error messages since it doesn't reconstruct the expressions 1:1.

Instead you should just return something like "Type mismatch. Expected: «type», got «type»" and the position attached to the AST. Message printer would use Megaparsec to construct an error message which would print the exact code that caused the error.